fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java

```diff
@@ -695,51 +713,50 @@ private void enforce(String resourceType, String resourceId, Resource resource,
-                        s.getResourceType().value().equals(resourceType))
-                .filter(s -> hasPermission(s.getPermission(), requiredPermission))
-                // Then group the scopes by their context type
-                .collect(Collectors.groupingBy(s -> s.getContextType()));
-
-        if (approvedScopeMap.containsKey(ContextType.SYSTEM)) {
-            if (log.isLoggable(Level.FINE)) {
-                log.fine(requiredPermission.value() + " permission for '" + resourceType + "/" + resourceId +
-                    "' is granted via scope " + approvedScopeMap.get(ContextType.SYSTEM));
+        for (ContextType context : List.of(ContextType.SYSTEM, ContextType.USER)) {
+            // For `system` and `user` scopes, we grant access to all resources of the requested type.
+            // Implementers that use these scopes are encouraged to layer on their own permissions model beyond this.
+            if (!grantedScopes.containsKey(context)) {
+                continue;
             }
-            return true;
-        }
 
-        if (approvedScopeMap.containsKey(ContextType.USER)) {
-            // For `user` scopes, we grant access to all resources of the requested type.
-            // Implementers that use these scopes are encouraged to layer on their own permissions model beyond this.
-            if (log.isLoggable(Level.FINE)) {
-                log.fine(requiredPermission.value() + " permission for '" + resourceType + "/" + resourceId +
-                    "' is granted via scope " + approvedScopeMap.get(ContextType.USER));
+            Optional<Scope> approvingScope = grantedScopes.get(context).stream()
+                    .filter(s -> s.getResourceType() == ResourceType.Value.RESOURCE || s.getResourceType().value().equals(resourceType))
+                    .filter(s -> hasPermission(s.getPermission(), requiredPermission))
+                    .findAny();
+            if (approvingScope.isPresent()) {
+                if (log.isLoggable(Level.FINE)) {
+                    log.fine(requiredPermission.value() + " permission for '" + resourceType + "/" + resourceId +
+                        "' is granted via scope " + approvingScope.get());
+                }
+                return true;
             }
-            return true;
         }
 
-        if (approvedScopeMap.containsKey(ContextType.PATIENT)) {
+        if (grantedScopes.containsKey(ContextType.PATIENT)) {
+            Optional<Scope> approvingScope = grantedScopes.get(ContextType.PATIENT).stream()
+                    .filter(s -> s.getResourceType() == ResourceType.Value.RESOURCE || s.getResourceType().value().equals(resourceType))
+                    .filter(s -> hasPermission(s.getPermission(), requiredPermission))
+                    .findAny();
+
             if (resource == null) {
                 // For `patient` scopes, if no resource was retrieved, which happens with history/search
```

Comment:
what about deleted resources? The resource value will be null for those too

Reply:
we don't call this from `afterDelete`, but we do call it from `afterHistory` and so I agree this is worth testing...I'll give it a try and maybe add a test

Reply:
this wasn't really from this PR, but instead from https://github.com/IBM/FHIR/issues/3423
however, i confirmed that, with those changes, we now reject a history interaction if any one of the resource versions in the history was a delete...i'll see if i can address that in this PR.  if not, i'll open a separate issue for it

Reply:
I made some changes to try addressing this one.  please have another look